File: android/app/src/main/kotlin/com/aer/notes_insa/WorkerStore.kt
package com.aer.notes_insa

import android.content.Context
import android.content.SharedPreferences
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import android.util.Base64
import android.util.Log
import java.security.KeyStore
import javax.crypto.Cipher
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec

/**
 * Encrypted key/value store shared between the Flutter app (which writes through
 * the MethodChannel) and [GradesBackgroundWorker] (which reads and updates it).
 *
 * Each value is sealed with an AES-256-GCM key held in the Android Keystore
 * (non-exportable, hardware-backed where available) and the ciphertext is kept
 * in a regular private SharedPreferences file. This replaces the deprecated
 * AndroidX EncryptedSharedPreferences (Jetpack Security), which flutter_secure_
 * storage itself dropped for the same reason.
 *
 * flutter_secure_storage 10.x stores values under its own cipher with prefixed
 * keys the worker cannot read, so the app mirrors the few secrets the worker
 * needs into this store on every credential change. Key names must stay in sync
 * with WorkerSyncService in lib/services/worker_sync_service.dart.
 */
object WorkerStore {
    private const val TAG = "WorkerStore"
    private const val FILE_NAME = "NotesInsaWorkerStoreV2"
    private const val LEGACY_FILE_NAME = "NotesInsaWorkerStore"
    private const val KEY_ALIAS = "NotesInsaWorkerStoreKey"
    private const val ANDROID_KEYSTORE = "AndroidKeyStore"
    private const val TRANSFORMATION = "AES/GCM/NoPadding"
    private const val GCM_TAG_BITS = 128
    private const val IV_BYTES = 12

    const val KEY_USERNAME = "username"
    const val KEY_PASSWORD = "password"
    const val KEY_OTP_SECRET = "otp_secret"
    const val KEY_CAS_SESSION = "cas_session"
    const val KEY_GRADES_JSON = "stored_grades_json"
    const val KEY_GRADES_UPDATED_AT = "stored_grades_updated_at"

    // TOTP step (floor(epochSeconds / 30)) most recently claimed by an
    // autoValidate caller, so the worker and the foreground don't submit the
    // same one-time code in the same step. See GradesBackgroundWorker.kt,
    // GradesBackgroundTask.swift, and grades_provider.dart.
    const val KEY_LAST_TOTP_STEP = "last_totp_step"

    // Deletes the abandoned pre-V2 EncryptedSharedPreferences file once per
    // process, so upgraded users who never log out don't keep it around.
    @Volatile
    private var legacyCleaned = false

    private fun cleanLegacyOnce(context: Context) {
        if (legacyCleaned) return
        legacyCleaned = true
        try {
            context.deleteSharedPreferences(LEGACY_FILE_NAME)
        } catch (e: Exception) {
            // Ignore; the legacy file is no longer read regardless.
        }
    }

    private fun prefs(context: Context): SharedPreferences {
        cleanLegacyOnce(context)
        return context.getSharedPreferences(FILE_NAME, Context.MODE_PRIVATE)
    }

    /**
     * Returns the Keystore AES key, generating it on first use. Synchronized so
     * a concurrent worker and foreground access cannot both generate (and thus
     * overwrite) the key on the very first use, which would leave earlier
     * ciphertext undecryptable.
     */
    @Synchronized
    private fun getOrCreateKey(): SecretKey {
        val keyStore = KeyStore.getInstance(ANDROID_KEYSTORE).apply { load(null) }
        (keyStore.getEntry(KEY_ALIAS, null) as? KeyStore.SecretKeyEntry)?.let {
            return it.secretKey
        }
        val generator = KeyGenerator.getInstance(
            KeyProperties.KEY_ALGORITHM_AES,
            ANDROID_KEYSTORE,
        )
        generator.init(
            KeyGenParameterSpec.Builder(
                KEY_ALIAS,
                KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT,
            )
                .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
                .setKeySize(256)
                .build(),
        )
        return generator.generateKey()
    }

    /** Seals [plaintext] as base64(iv || ciphertext) under the Keystore key. */
    private fun encrypt(plaintext: String): String {
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, getOrCreateKey())
        val iv = cipher.iv
        val ciphertext = cipher.doFinal(plaintext.toByteArray(Charsets.UTF_8))
        val combined = ByteArray(iv.size + ciphertext.size)
        System.arraycopy(iv, 0, combined, 0, iv.size)
        System.arraycopy(ciphertext, 0, combined, iv.size, ciphertext.size)
        return Base64.encodeToString(combined, Base64.NO_WRAP)
    }

    /** Reverses [encrypt], returning null if the value cannot be decrypted. */
    private fun decrypt(encoded: String): String? = try {
        val combined = Base64.decode(encoded, Base64.NO_WRAP)
        val iv = combined.copyOfRange(0, IV_BYTES)
        val ciphertext = combined.copyOfRange(IV_BYTES, combined.size)
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(
            Cipher.DECRYPT_MODE,
            getOrCreateKey(),
            GCMParameterSpec(GCM_TAG_BITS, iv),
        )
        String(cipher.doFinal(ciphertext), Charsets.UTF_8)
    } catch (e: Exception) {
        Log.e(TAG, "Failed to decrypt a stored value")
        null
    }

    /**
     * Reads and decrypts the requested keys. Returns null if the store cannot be
     * opened; keys with no stored value (or that fail to decrypt) come back as
     * null entries.
     */
    fun read(context: Context, keys: List<String>): Map<String, String?>? = try {
        val prefs = prefs(context)
        keys.associateWith { key -> prefs.getString(key, null)?.let(::decrypt) }
    } catch (e: Exception) {
        Log.e(TAG, "Failed to read worker store")
        null
    }

    /**
     * Encrypts and writes the provided keys. A null value removes that key. Keys
     * absent from [values] are left untouched, so callers can sync a subset.
     */
    fun write(context: Context, values: Map<String, String?>) {
        try {
            val editor = prefs(context).edit()
            for ((key, value) in values) {
                if (value == null) editor.remove(key) else editor.putString(key, encrypt(value))
            }
            editor.apply()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to write worker store")
        }
    }

    /** Clears all stored data (called on logout), durably reporting failure. */
    fun clearAll(context: Context) {
        if (!prefs(context).edit().clear().commit()) {
            throw IllegalStateException("Failed to clear worker store")
        }
    }
}

File: ios/Runner/WorkerStore.swift
import Foundation
import Security

/// iOS counterpart of Android's `WorkerStore.kt`.
///
/// On Android the worker store is a dedicated AndroidX EncryptedSharedPreferences
/// file shared between the Flutter app and the background WorkManager worker.
/// On iOS the background task (`GradesBackgroundTask`) runs *in the app's own
/// process*, so a plain Keychain item (no App Group / sharing entitlement
/// required) is sufficient. We use a distinct Keychain service so these mirrored
/// secrets stay separate from `flutter_secure_storage`'s own items.
///
/// Items are stored with `kSecAttrAccessibleAfterFirstUnlock` so the background
/// task can read them while the device is locked (after the first unlock since
/// boot).
///
/// Key names MUST stay in sync with `WorkerSyncService` in
/// lib/services/worker_sync_service.dart and with `WorkerStore.kt`.
enum WorkerStore {
    private static let service = "NotesInsaWorkerStore"

    static let keyUsername = "username"
    static let keyPassword = "password"
    static let keyOtpSecret = "otp_secret"
    static let keyCasSession = "cas_session"
    static let keyGradesJson = "stored_grades_json"
    static let keyGradesUpdatedAt = "stored_grades_updated_at"

    // MARK: - Public API (mirrors WorkerStore.kt: read / write / clearAll)

    /// Reads the requested keys. Keys with no stored value come back as `nil`.
    static func read(keys: [String]) -> [String: String?] {
        var result: [String: String?] = [:]
        for key in keys {
            // updateValue keeps the key present with a nil value when nothing is
            // stored (matching Kotlin's `associateWith`). Plain `result[key] =
            // nil` would instead REMOVE the key, so don't simplify this.
            result.updateValue(get(key), forKey: key)
        }
        return result
    }

    /// Writes the provided keys. A `nil` value removes that key. Keys absent
    /// from `values` are left untouched, so callers can sync a subset.
    static func write(values: [String: String?]) {
        for (key, value) in values {
            if let value = value {
                set(key, value)
            } else {
                remove(key)
            }
        }
    }

    /// Clears every stored value (called on logout).
    static func clearAll() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
        ]
        SecItemDelete(query as CFDictionary)
    }

    // MARK: - Single-item helpers

    static func get(_ key: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        let status = SecItemCopyMatching(query as CFDictionary, &item)
        guard status == errSecSuccess,
              let data = item as? Data,
              let value = String(data: data, encoding: .utf8) else {
            return nil
        }
        return value
    }

    private static func set(_ key: String, _ value: String) {
        let data = Data(value.utf8)
        let base: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
        // Try to update an existing item first; insert if it doesn't exist.
        let update: [String: Any] = [
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock,
        ]
        let status = SecItemUpdate(base as CFDictionary, update as CFDictionary)
        if status == errSecItemNotFound {
            var insert = base
            insert[kSecValueData as String] = data
            insert[kSecAttrAccessible as String] = kSecAttrAccessibleAfterFirstUnlock
            SecItemAdd(insert as CFDictionary, nil)
        }
    }

    private static func remove(_ key: String) {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: key,
        ]
        SecItemDelete(query as CFDictionary)
    }
}
